File: mobile/test/domain/dates_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:my_finance/domain/dates.dart';

void main() {
  group('parseIso / toIso', () {
    test('ida y vuelta', () {
      expect(parseIso('2026-09-18'), const Ymd(2026, 9, 18));
      expect(toIso(const Ymd(2026, 9, 18)), '2026-09-18');
    });

    test('rellena con ceros', () {
      expect(toIso(const Ymd(2026, 1, 5)), '2026-01-05');
    });

    test('rechaza basura', () {
      expect(() => parseIso('18/09/2026'), throwsFormatException);
      expect(() => parseIso('2026-09'), throwsFormatException);
    });
  });

  group('daysInMonth', () {
    test('meses normales', () {
      expect(daysInMonth(2026, 1), 31);
      expect(daysInMonth(2026, 4), 30);
    });

    test('febrero bisiesto y no bisiesto', () {
      expect(daysInMonth(2024, 2), 29);
      expect(daysInMonth(2026, 2), 28);
      expect(daysInMonth(2000, 2), 29);
      expect(daysInMonth(1900, 2), 28);
    });
  });

  group('clampDay', () {
    test('nunca produce un día inexistente', () {
      expect(clampDay(2026, 2, 31), 28);
      expect(clampDay(2024, 2, 31), 29);
      expect(clampDay(2026, 1, 31), 31);
      expect(clampDay(2026, 1, 0), 1);
    });
  });

  group('shiftMonth', () {
    test('cruza el año hacia adelante y hacia atrás', () {
      expect(shiftMonth(2026, 12, 1), (y: 2027, m: 1));
      expect(shiftMonth(2026, 1, -1), (y: 2025, m: 12));
      expect(shiftMonth(2026, 6, 0), (y: 2026, m: 6));
      expect(shiftMonth(2026, 1, -13), (y: 2024, m: 12));
    });
  });

  group('addDays', () {
    test('cruza mes y año', () {
      expect(addDays(const Ymd(2026, 1, 31), 1), const Ymd(2026, 2, 1));
      expect(addDays(const Ymd(2026, 12, 31), 1), const Ymd(2027, 1, 1));
      expect(addDays(const Ymd(2026, 3, 1), -1), const Ymd(2026, 2, 28));
    });
  });

  test('compareIso ordena como string ISO', () {
    expect(compareIso('2026-01-01', '2026-01-02'), -1);
    expect(compareIso('2026-01-02', '2026-01-01'), 1);
    expect(compareIso('2026-01-01', '2026-01-01'), 0);
  });

  test('weekdayOf usa 0=domingo, como getUTCDay', () {
    expect(weekdayOf(const Ymd(2026, 9, 20)), 0); // domingo
    expect(weekdayOf(const Ymd(2026, 9, 18)), 5); // viernes
  });

  test('todayIso usa la fecha local del dispositivo', () {
    expect(todayIso(DateTime(2026, 9, 18, 23, 30)), '2026-09-18');
  });
}

File: src/domain/dates.test.ts
import { describe, expect, it } from 'vitest';
import { addDays, clampDay, daysInMonth, parseISO, shiftMonth, toISO } from './dates';

describe('daysInMonth', () => {
  it('reconoce años bisiestos', () => {
    expect(daysInMonth(2024, 2)).toBe(29); // bisiesto
    expect(daysInMonth(2026, 2)).toBe(28); // no bisiesto
    expect(daysInMonth(2026, 1)).toBe(31);
    expect(daysInMonth(2026, 4)).toBe(30);
  });
});

describe('clampDay', () => {
  it('nunca produce un dia inexistente', () => {
    expect(clampDay(2026, 2, 31)).toBe(28);
    expect(clampDay(2024, 2, 31)).toBe(29);
    expect(clampDay(2026, 4, 31)).toBe(30);
    expect(clampDay(2026, 1, 15)).toBe(15);
  });
});

describe('shiftMonth', () => {
  it('cruza de año hacia adelante', () => {
    expect(shiftMonth(2026, 12, 1)).toEqual({ y: 2027, m: 1 });
  });
  it('cruza de año hacia atras', () => {
    expect(shiftMonth(2027, 1, -1)).toEqual({ y: 2026, m: 12 });
  });
  it('avanza varios meses de una vez', () => {
    expect(shiftMonth(2026, 10, 4)).toEqual({ y: 2027, m: 2 });
  });
});

describe('addDays / toISO / parseISO', () => {
  it('cruza de mes correctamente', () => {
    expect(toISO(addDays(parseISO('2026-01-31'), 1))).toBe('2026-02-01');
  });
  it('cruza de año correctamente', () => {
    expect(toISO(addDays(parseISO('2026-12-31'), 1))).toBe('2027-01-01');
  });
  it('resta dias sin producir dia 0', () => {
    expect(toISO(addDays(parseISO('2026-03-01'), -1))).toBe('2026-02-28');
  });
  it('parseISO/toISO son inversas', () => {
    expect(toISO(parseISO('2026-09-17'))).toBe('2026-09-17');
  });
});
